### ISSUE TITLE
Do Ical better

### ISSUE THREAD

**Issue by shawfactor** [GatherPress/gatherpress#603]
### Is your enhancement related to a problem? Please describe.

Currently plugin supports ical which is great and essential for all event plugins

However the support is in the form of a data uri. This is FAR from ideal and should be reconsidered. I would suggest a much better approach would unique url that respsents the ical version of the events.

Possible utiising the wordpress custom feed functionality e.g. 

for the singular version: https://mysite.com/event/test-event/feed/ical

 for index of events: https://mysite.com/future-events/feed/ical/

or utilising a get parameter e.g.

for the singular version: https://mysite.com/event/test-event/?ical=1

 for index of events: https://mysite.com/future-events/?ical=1

Bonus points if this scheme suppoorted generating an ical calendar based on a taxonomy e.g. evets with a certain topic. 

This approach would have the major adavantage of allowing devices like iphones to subscribe to the calendar of events that would update dyamically. Rather than adding an event to ones calendar that are static and may well end up out of date. More broader is would also be a restful approach.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by carstingaxion** [comment]
Following the outlined idea by @shawfactor would also fix #704, because the 🐛 mentioned is coming from the used data-uri, too.

**Comment by carstingaxion** [comment]
GatherPress should probably replace the existing (and default) `/feeds/` endpoints with new `/ical/` endpoints for:

- `gatherpress_events` post type
- `gatherpress_venues` post type
- `gatherpress_topics` taxonomy

While this seems identical, the endpoints will have slightly different purposes per data type.

- The `gatherpress_events` ical allows to subscribe to all events of that site and needs to be added on the post type level.
- The `gatherpress_venues` ical allows to all events of a particular venue. Therefore it must be added on the post level.
- The `gatherpress_topics` ical returns all events related to one particular topic term. The endpoint needs to be added to the term level.

---

- https://developer.wordpress.org/apis/rewrite/
- https://make.wordpress.org/plugins/2012/06/07/rewrite-endpoints-api/
- https://developer.wordpress.org/reference/functions/add_rewrite_endpoint/
- https://developer.wordpress.org/reference/functions/add_feed/

**Comment by shawfactor** [comment]
Replace or add?

I’d suggest the best approach is to add a custom feed (in iCal format)

Eg events would have

https://mysite.com/events/ (archive)
https://mysite.com/events/feed/ (rss feed)
https://mysite.com/events/feed/atom/ (atom feed)
https://mysite.com/events/feed/rdf/ (rdf feed)

Then adding

https://mysite.com/events/feed/ical/ (iCal feed)

RSS, atom, and rdf come built in with Wordpress and I don’t see the need to
get rid of them.

Feeds also play nice with taxonomies so being able to subscribe to just
events based on a single topic would be very beneficial

I’ve actually written a plugin that does this as I wanted this
functionality on my own sites btw

**Comment by carstingaxion** [comment]
> Replace or add?

Originally I meant _replace_, but it may make sense to rethink this.

> I’d suggest the best approach is to add a custom feed (in iCal format)

Yes, absolutely.
> 
> Eg events would have
> 
> https://mysite.com/events/ (archive)
> https://mysite.com/events/feed/ (rss feed)
> https://mysite.com/events/feed/atom/ (atom feed)
> https://mysite.com/events/feed/rdf/ (rdf feed)
> 
> Then adding
> https://mysite.com/events/feed/ical/ (iCal feed)
> 
> RSS, atom, and rdf come built in with Wordpress and I don’t see the need to
> 
> get rid of them.

But using the default feeds functionality _as is_, would require at least some adjustments to the rendered feeds, as it would not be that helpful to have events ordered by their publication date for example.

> Feeds also play nice with taxonomies so being able to subscribe to just
> 
> events based on a single topic would be very beneficial

Agree. That’s what I meant with adding iCal support to the topics taxonomy.

> I’ve actually written a plugin that does this as I wanted this
> 
> functionality on my own sites btw

Cool!
Feel free to share your work with us to help moving GatherPress forward, otherwise we would need to implement this on our own.

**Comment by MervinHernandez** [comment]
QUEUED to Discuss October 25
- ?? How is this related to #704?
- ?? Has anyone submitted any code to address this at all?

**Comment by MervinHernandez** [comment]
Hi @carstingaxion  checking on where this stands with you -- and whether we can pin it for 0.32?

**Comment by MervinHernandez** [comment]
Hi @carstingaxion checking once more on the status and plausibility of marking this into 0.32?

**Comment by carstingaxion** [comment]
Hello, I tend to say yes, because 75% are already merged and the last missing piece is awaiting review.

Please take a look at #... and add your review to move this forward.

**Comment by MervinHernandez** [comment]
Thank you @carstingaxion // acknowledging this is ready for review. 
I will schedule actions into reviewing this as soon as possible.

**Comment by JordanPak** [comment]
I'ma check this out @mauteri

**Comment by carstingaxion** [comment]
Hello @JordanPak have you found the time to check this out, what do you think?

**Comment by JordanPak** [comment]
@carstingaxion unfortunately I was caught up with moving to a new home, then having to start a job search. I may be able to look at it soon if the market continues to be a mess though 🪦 😬

**Comment by heiglandreas** [comment]
What is missing to get this done and released? This is the one thing blocking us (The PHP-Usergropu in Frankfurt at https://phpugffm.de) with GatherPress and makes us consider to move back to the EventManager Plugin (as that has ical-feeds out of the box).

Having an ical feed for a website that manages events is a vital requirement for various reasons. Visibility in non-activity-pub websites is only one but not a small one.

If there is anything I can do to help this ship, feel free to reply.

**Comment by mauteri** [comment]
Really appreciate the offer, @heiglandreas. Being candid: this has bounced through a few hands and I'm not the best fit to drive it home. If you'd like to take it on, I'd genuinely love that. I can help with reviews and unblocking, but I think this needs an owner who isn't me to ship it the right way. @carstingaxion has been close to this one and may want to be part of moving it forward.

**Comment by mauteri** [comment]
@heiglandreas we'll be prioritizing this now. I'll add it to 0.34.0 release and see if we can squeeze it in there.

**Comment by carstingaxion** [comment]
I will prepare an updated PR .

**Comment by mauteri** [comment]
@carstingaxion @jmarx if you can carve out some time to look at this and share feedback this week, that'd be great. Trying to keep 0.34.0 moving. Once it lands and gets tested I'd like to cut a beta and ship after some more testing. No pressure if the week's tight. I'm happy to move it to a later milestone if needed, but considering recent feedback, it would be great to get this in 0.34.0. Thx!

**Comment by carstingaxion** [comment]
Im not sure what kind of feedback you're looking for @mauteri, but what I'd like to have, is a minimal consensus on the plan, before I start wrangling a lot of code around for the third time.

---

### Goal

Get subscribable calendar feeds setup for the following routes

1. `example.org/feed/ical`
2. `example.org/event/feed/ical`
3. `example.org/event/my-event/ical`
4. `example.org/venue/my-location/feed/ical`
5. `example.org/topic/my-topic/feed/ical`
6. `example.org/ANY-TAX-OF-EVENTS/my-term/feed/ical`

### Implementation 

*Most of this is already done in #955 and will be ported into a new PR if that aligns with GatherPress' goals. All the details are already here: [docs/developer/custom-url-endpoints](https://github.com/GatherPress/gatherpress/tree/develop/docs/developer/custom-url-endpoints)*

1. A new "endpoint registry" allows to create any custom endpoint to be registered for either 

   a) the whole site
   b) a post type
   c) a post
   d) a taxonomy term

   What gets registered with each endpoint is the endpoint-slug together with a template-path string to the file where the rendering happens. 

2. The New files under `includes/templates/ical/*.php` can easily overwritten by any theme.

3. The registry is made as agnostic as possible so future GatherPress or extension developers can utilise the methods to be able to create routes like

   a) `example.org/event/my-event/qr-code`
   b) `example.org/venue/my-venue/directions`

4. With the recent architectural changes to use post_type_supports more widely, the registry here should be directed via the same mechanisms if possible.

5. The actual ics generation in my initial implementation is mostly similar to what is currently in GatherPress core and should be merged. Additionally creating ics data for a loop of posts will be made possible to allow to render the new post_type- and taxonomy-level endpoints.
### Needs feedback

- [ ] My last attempt at #955 included changing the "Add to calendar" links to human readable slugs like `...my-event/googlecalendar` to avoid that long uris and have them easily available in themes etc. 
- [ ] Initially I wrote the docs first to bring everybody on track about the implementation. The current ics-endpoint handling differs from what I suggested there, especially about not using wp_... 
- [ ] To be able to use the new endpoints in FSE, either https://github.com/WordPress/gutenberg/issues/65351 OR https://github.com/carstingaxion/gatherpress-feed-buttons would be needed
- [ ] Enabling both endpoints on `example.org/feed/ical` and on 
`example.org/event/feed/ical` will result in redundant data and maybe even generate a duplicate content issue for SEO. *We should decide which of both should be the enabled default!*
- [x] The original PR didn't include any tests, but @jmarx already jumped in (on slack) to offer help, *which is great, thx!*
- [ ] The implicit suggestion by @heiglandreas to only query for upcoming events makes sense, wasn’t part of the former PR and should be implemented 
- [ ] The real suggestion by @heiglandreas to setup Content-Negotiation routes makes sense too, also wasn’t part of the former PR and should be implemented, too

**UPDATE:** Added `/feed` into most of the routes, which I forgot in the first try

**Comment by heiglandreas** [comment]
Just for clarification:

Do I get this right?

* example.org/ical - ICS feed of ALL events (only future events?)
* example.org/event/ical - ICS feed of ... just one event?
* example.org/venue/my-location/ical - ICS feed of ALL events at a certain venue (only future events?)
* example.org/topic/my-topic/ical - ICS feed of ALL events of a certain topic (only future events?)
* example.org/ANY-TAX-OF-EVENTS/my-term/ical - ICS feed of ALL events of the taxonomy "my-term" (only future events?)

Would that not be `example.org/event/my-event/ical` for just the one event?

Also: If that's not too much trouble it would be awesome to use Content-Negotiation to allow 

```
GET /
Host: example.org
Accept: text/calendar
```

and to have that either redirect to `example.org/ical` or to directly fetch the respective ICS file.

**Comment by mauteri** [comment]
@carstingaxion Plan looks right. Go ahead with the port. On your four open questions:

Add-to-Calendar slugs: defer. Separate enhancement, shouldn't block this.
Doc drift: update docs to match as-built during the port.
FSE block: ship the URL endpoints first. Editor block is a follow-up; existing add-to-calendar block holds the line.
/feed/ical vs /event/feed/ical: ship both. Site-wide merges all event-supporting post types (productions, events, etc.), the post-type one is just events. SEO duplicate-content concern doesn't apply to text/calendar; search engines don't index ICS.

Andreas's content-negotiation idea is a nice follow-up but not core. Calendar apps need an explicit URL anyway.

@heiglandreas on future-vs-past: feed routes (1, 2, 4, 5, 6) default to future-only, since past events are noise for a subscription. Single event (route 3) is one VEVENT either way. Filterable to override.

Three operational gotchas worth designing for during the port (shouldn't slow you down):

Caching. Outlook and Google poll subscribed feeds frequently. HTTP cache headers (Cache-Control: max-age, ETag) plus object-cache backing, busted on event save.
Private events. Public is the right default, but private events need a token-protected URL story (Google Calendar's /private/[token]/basic.ics shape). OK as a follow-up, just acknowledge the path.
Recurrence forward-compat. No recurring events today, but when they land the ICS needs RRULE. Just don't paint into a corner that makes that painful.

**Comment by mauteri** [comment]
Work is being done here: https://github.com/GatherPress/gatherpress/pull/955

**Comment by carstingaxion** [comment]
Now that the major part is merged, let not forget about all the good ideas in the last comments.

@mauteri @heiglandreas Do you mind creating fresh issues with the ideas of your last recent comments from this thread. Just to make sure, we don't forget! Thanks.

**Comment by mauteri** [comment]
@carstingaxion ✅ Filed five follow-up issues from the last round of comments on this thread:

- **#1682** — Add HTTP cache headers + object-cache backing for ICS feeds (Cache-Control / ETag / Last-Modified, busted on event save). From [my comment](https://github.com/GatherPress/gatherpress/issues/603#issuecomment-4364756998).
- **#1683** — Add token-protected ICS URLs for private events (Google Calendar `/private/[token]/basic.ics` shape). From [my comment](https://github.com/GatherPress/gatherpress/issues/603#issuecomment-4364756998).
- **#1684** — Reserve ICS shape for future RRULE support when recurring events (#80) land. From [my comment](https://github.com/GatherPress/gatherpress/issues/603#issuecomment-4364756998).
- **#1685** — Support `Accept: text/calendar` content-negotiation for event URLs. From [@heiglandreas's comment](https://github.com/GatherPress/gatherpress/issues/603#issuecomment-4377380791).
- **#1686** — Add FSE/editor block for ICS subscribe links (vendor in `gatherpress-feed-buttons` or upstream `gutenberg#65351`). From [your checklist](https://github.com/GatherPress/gatherpress/issues/603#issuecomment-4377157417).

All five are 0.35.0+ — not landing in 0.34.0.

Also separately filed **#1687** from your Slack note about generalizing the `venue-detail` block to work with any shadow-source post type (Tour, Production, Season). Also 0.35.0+.

**Comment by mauteri** [comment]
Spun out the event-query block's contextual-filter generalization as **#1688**, sibling of #1687.

Per @carstingaxion's note:

> "It's the combination of that details block (variation) and the toggle on the query block, that makes this whole logic shine to the stars."

Updated #1687 to cross-link #1688 and call out that combination explicitly — independently each is useful, together they unlock the full shadow-source extensibility story (single page template for a tour / production / season / venue, no per-type code).
